src/providers/BackgroundReplacementProvider/index.tsx

```diff
@@ -0,0 +1,186 @@
+import { BaseSdkProps } from '../../components/sdk/Base';
+
+interface Props extends BaseSdkProps {
+  /** The spec defines the assets that will be used for adding background replacement to a frame. For more information, refer to
+   * [Amazon Chime JS SDK Background Filter Guide](https://github.com/aws/amazon-chime-sdk-js/blob/main/guides/15_Background_Filter_Video_Processor.md#adding-a-background-filter-to-your-application). */
+  spec?: BackgroundFilterSpec;
+  /* A set of options that can be supplied when creating a background replacement video frame processor such as the background replacement image blob. For more information, refer to
+   * [Amazon Chime JS SDK Background Filter Guide](https://github.com/aws/amazon-chime-sdk-js/blob/main/guides/15_Background_Filter_Video_Processor.md#adding-a-background-filter-to-your-application). */
+  options?: BackgroundReplacementOptions;
+}
+
+interface BackgroundReplacementProviderState {
+  createBackgroundReplacementDevice: (
+    device: Device
+  ) => Promise<DefaultVideoTransformDevice>;
+  isBackgroundReplacementSupported: boolean | undefined;
+}
+
+const BackgroundReplacementProviderContext =
+  createContext<BackgroundReplacementProviderState | undefined>(undefined);
+
+const BackgroundReplacementProvider: FC<Props> = ( { spec, options, children }) => {
+  const [isBackgroundReplacementSupported, setIsBackgroundReplacementSupported] = useState<boolean | undefined>(undefined);
+  const [processor, setProcessor] = useState<VideoFrameProcessor | undefined>(undefined);
+  const processorState = useRef<VideoFrameProcessor | undefined>();
+  processorState.current = processor;
+
+  const replacementSpec = useMemoCompare(
+    spec,
+    (
+      prev: BackgroundFilterSpec | undefined,
+      next: BackgroundFilterSpec | undefined
+    ): boolean => {
+      if ((Object.is(prev, next)) || isPrevNextUndefined(prev, next)) {
+        return true;
+      }
+      return false;
+    }
+  );
+
+  const replacementOptions = useMemoCompare(
+    options,
+    (
+      prev: BackgroundReplacementOptions | undefined,
+      next: BackgroundReplacementOptions | undefined
+    ): boolean => {
+      if ( isPrevNextUndefined(prev, next) || (prev?.imageBlob?.size == next?.imageBlob?.size && Object.is(prev?.filterCPUUtilization, next?.filterCPUUtilization))) {
+        return true;
+      }
+      return false;
+    }
+  );
+  
+  useEffect(() =>{
+    console.log('Specs or options were changed. Re-initializing background replacement processor.');
+    initializeBackgroundReplacement();
+    return () => {
+      console.log(`Destroying background replacement processor.`);
+      processor?.destroy();
+    };
+  }, [replacementSpec, replacementOptions]);
+
+
+  let currentProcessorPromise:
+  | Promise<BackgroundReplacementProcessor| undefined>
+  | undefined;
+
+  async function initializeBackgroundReplacement(): Promise<BackgroundReplacementProcessor | undefined> {
```

Comment:
The reason I added the global promise was in the case that Props are changed which trigger to re-creating the processor while another promise to create a different processor with the old spec is ongoin by createBackgroundReplacementDevice. This way createBackgroundReplacementDevice will wait for the most recent promise of creating a processor.

Reply:
Thanks for the explanation.

1. Suppose that `createBackgroundReplacementDevice` waits for **PromiseA** to resolve. **PromiseA** is `currentProcessorPromise`.
2. Before **PromiseA** is resolved, a provider receives new props, calling `initializeBackgroundReplacement`. **PromiseB** is now `currentProcessorPromise`.
3. Step 1 completes. Since `createBackgroundReplacementDevice` waits for **PromiseA**, it will still get **PromiseA**.

> This way createBackgroundReplacementDevice will wait for the most recent promise of creating a processor.

At step 3, I think `createBackgroundReplacementDevice` will still wait for the original promise **PromiseA**.

Reply:
fixed